### PR TITLE
Move structures from `ty/mod.rs` into their own modules

### PR DESCRIPTION
**Pull request by WaffleLapkin**
Revival of #111208, because github refuses to work correctly.

r? @compiler-errors 
cc @cjgillot 

----

I've implemented most of the suggestions from #111208's reviews.

Exceptions:
- `ToPredicate` and `InstantiatedPredicates`
  - They are both not part of the predicate representation, but rather related things
  - IMO they make more sense as submodules
- One-off queries stuff
  - I'm not sure where to put them
  - An option could be `ty::query` (*and* move the macros in like `ty::query::macros`), but idk

I'd also like to split adt stuff at least a little, for example
```rust
mod adt {
    mod field {}
    mod variant {}
}
```

### CODE REVIEW

File: compiler/rustc_middle/src/ty/region.rs [hunk 1]
```diff
@@ -0,0 +1,3 @@
+use crate::ty::{BoundRegion, Placeholder};
```
Comment (compiler-errors) [inline compiler/rustc_middle/src/ty/region.rs:1]: You should just move it into sty with the rest of the region variants, then later if you want to break up sty, all the region variants can be moved into a file here. This file in its current state is kinda misleading, lmao

File: compiler/rustc_middle/src/ty/predicate/to_predicate.rs [hunk 2]
```diff
@@ -0,0 +1,103 @@
+use crate::ty::{
```
Comment (compiler-errors) [inline compiler/rustc_middle/src/ty/predicate/to_predicate.rs:1]: I still think that this should live in predicates. These are essentially just `Into` implementations for `Predicate`, and it's easier to just have them live with the types they're defined on.

File: compiler/rustc_middle/src/ty/destructured_const.rs [hunk 3]
```diff
@@ -0,0 +1,8 @@
+use crate::ty::{self, VariantIdx};
```
Comment (compiler-errors) [inline compiler/rustc_middle/src/ty/destructured_const.rs:1]: I think this probably should just live with the rest of the consts..

**Review by compiler-errors — COMMENTED** [review]
Maybe let's see what @cjgillot thinks. Also this should be squashed at the end -- it doesn't necessarily need to be 1 commit, but also I think 40 commits is a bit excessive.

File: compiler/rustc_middle/src/ty/predicate/to_predicate.rs (hunk 2 above)
Comment (BoxyUwU) [reply compiler/rustc_middle/src/ty/predicate/to_predicate.rs:1]: I feel like we ought to just not have any of the `predicate/` files and isntead put it all in `predicate.rs` there's barely anything in the submodules and they would fit in fine in `predicate.rs`

File: compiler/rustc_middle/src/ty/destructor.rs [hunk 4]
```diff
@@ -0,0 +1,10 @@
+use rustc_hir as hir;
+use rustc_hir::def_id::DefId;
+
+#[derive(Copy, Clone, Debug, HashStable, Encodable, Decodable)]
+pub struct Destructor {
```
Comment (cjgillot) [inline compiler/rustc_middle/src/ty/destructor.rs:5]: In `ty::adt`?

File: compiler/rustc_middle/src/ty/predicate/crate_predicates_map.rs [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+use crate::ty::Clause;
+
+/// The crate outlives map is computed during typeck and contains the
+/// outlives of every item in the local crate. You should not use it
+/// directly, because to do so will make your pass dependent on the
+/// HIR of every item in the local crate. Instead, use
+/// `tcx.inferred_outlives_of()` to get the outlives for a *particular*
+/// item.
+#[derive(HashStable, Debug)]
+pub struct CratePredicatesMap<'tcx> {
```
Comment (cjgillot) [inline compiler/rustc_middle/src/ty/predicate/crate_predicates_map.rs:14]: In `ty::predicate`?

File: compiler/rustc_middle/src/ty/placeholder.rs [hunk 6]
```diff
@@ -0,0 +1,12 @@
+use crate::ty::UniverseIndex;
+
+/// The "placeholder index" fully defines a placeholder region, type, or const. Placeholders are
+/// identified by both a universe, as well as a name residing within that universe. Distinct bound
+/// regions/types/consts within the same universe simply have an unknown relationship to one
+/// another.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
+#[derive(HashStable, TyEncodable, TyDecodable)]
+pub struct Placeholder<T> {
```
Comment (cjgillot) [inline compiler/rustc_middle/src/ty/placeholder.rs:9]: In `rustc_type_ir`?

File: compiler/rustc_middle/src/ty/predicate/instantiated_predicates.rs [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+/// Example:
+/// ```ignore (illustrative)
+/// struct Foo<T, U: Bar<T>> { ... }
+/// ```
+/// Here, the `GenericPredicates` for `Foo` would contain a list of bounds like
+/// `[[], [U:Bar<T>]]`. Now if there were some particular reference
+/// like `Foo<isize,usize>`, then the `InstantiatedPredicates` would be `[[],
+/// [usize:Bar<isize>]]`.
+#[derive(Clone, Debug, TypeFoldable, TypeVisitable)]
+pub struct InstantiatedPredicates<'tcx> {
```
Comment (cjgillot) [inline compiler/rustc_middle/src/ty/predicate/instantiated_predicates.rs:25]: Should this go in `ty::predicate` too?

File: compiler/rustc_middle/src/ty/ty_.rs [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+/// Use this rather than `TyKind`, whenever possible.
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, HashStable)]
+#[rustc_diagnostic_item = "Ty"]
+#[rustc_pass_by_value]
+pub struct Ty<'tcx>(pub(super) Interned<'tcx, WithCachedTypeInfo<TyKind<'tcx>>>);
+
+pub type PlaceholderType = Placeholder<BoundTy>;
+
+#[derive(Debug, Default, Copy, Clone)]
+pub struct InferVarInfo {
```
Comment (cjgillot) [inline compiler/rustc_middle/src/ty/ty_.rs:12]: This can be moved to `rustc_hir_typeck`.

File: compiler/rustc_middle/src/ty/ty_.rs [hunk 9]
```diff
@@ -0,0 +1,24 @@
+use crate::ty::{BoundTy, Interned, Placeholder, TyKind, WithCachedTypeInfo};
+
+/// Use this rather than `TyKind`, whenever possible.
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, HashStable)]
+#[rustc_diagnostic_item = "Ty"]
+#[rustc_pass_by_value]
+pub struct Ty<'tcx>(pub(super) Interned<'tcx, WithCachedTypeInfo<TyKind<'tcx>>>);
+
+pub type PlaceholderType = Placeholder<BoundTy>;
```
Comment (cjgillot) [inline compiler/rustc_middle/src/ty/ty_.rs:9]: Those should be in sty until you get to break it.

File: compiler/rustc_middle/src/ty/predicate/to_predicate.rs (hunk 2 above)
Comment (WaffleLapkin) [reply compiler/rustc_middle/src/ty/predicate/to_predicate.rs:1]: @compiler-errors Most of those are not implemented for `Predicate`s though, so I'm not sure I follow...

@BoxyUwU I don't agree, having the same points as before: adding more entities to the same file makes it harder to add something there (it's easy to mess with the order) and harder to navigate... I can understand the reasoning behind having all predicate parts in a single file (after all, it's essentially a single thing split into multiple types), but I can't see the point in merging `ToPredicate`, `InstantiatedPredicates` and `CratePredicatesMap` there too.

**Comment by apiraino** [comment]
I see a few open comments so I think this can be switched to waiting on the author

@rustbot author

**Comment by WaffleLapkin** [comment]
This is not going anywhere in this state. I'll try to do smaller PRs that only move items of a certain _theme_, so that it's actually possible to review and do the changes (this is basically impossible to rebase).

**Comment by WaffleLapkin** [comment]
Status update: I did not in fact do smaller PRs. I still think that giant `mod.rs` is not the best, but oh well. Maybe I'll find time some day

**Pull request closed without merging**